gcp/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryClientImpl.java

```diff
@@ -0,0 +1,481 @@
+    validateTable(table); // Ensure it is an Iceberg table supported by the catalog
+
+    ExternalCatalogTableOptions newExternalCatalogTableOptions =
+        new ExternalCatalogTableOptions()
+            .setStorageDescriptor(table.getExternalCatalogTableOptions().getStorageDescriptor())
+            .setConnectionId(table.getExternalCatalogTableOptions().getConnectionId())
+            .setParameters(table.getExternalCatalogTableOptions().getParameters());
+    Table patch =
+        new Table()
+            .setExternalCatalogTableOptions(newExternalCatalogTableOptions)
+            // TODO(b/341933455): Update this once the server side accepts schema.
+            // Must set the schema as null for using schema auto-detect
+            .setSchema(Data.nullOf(TableSchema.class));
+
+    try {
+      HttpResponse response =
+          client
+              .tables()
+              .patch(
+                  tableReference.getProjectId(),
+                  tableReference.getDatasetId(),
+                  tableReference.getTableId(),
+                  patch)
+              .setRequestHeaders(new HttpHeaders().setIfMatch(table.getEtag()))
+              .executeUnparsed();
+
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        String responseString = response.parseAsString();
+        if (responseString.toLowerCase(Locale.ROOT).contains("not found: connection")) {
+          throw new BadRequestException(responseString);
+        }
+        throw new NoSuchTableException(response.getStatusMessage());
+      }
+      return convertExceptionIfUnsuccessful(response).parseAs(Table.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @Override
+  @SuppressWarnings("FormatStringAnnotation")
+  public Table renameTable(TableReference tableToRename, String newTableId) {
+    Table table = getTable(tableToRename); // Verify table first
+    Table patch =
+        new Table()
+            .setTableReference(
+                new TableReference()
+                    .setProjectId(table.getTableReference().getProjectId())
+                    .setDatasetId(table.getTableReference().getDatasetId())
+                    .setTableId(newTableId));
+
+    try {
+      HttpResponse response =
+          client
+              .tables()
+              .patch(
+                  tableToRename.getProjectId(),
+                  tableToRename.getDatasetId(),
+                  tableToRename.getTableId(),
+                  patch)
+              .setRequestHeaders(new HttpHeaders().setIfMatch(table.getEtag()))
+              .executeUnparsed();
+
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchTableException(response.getStatusMessage());
+      }
+      return convertExceptionIfUnsuccessful(response).parseAs(Table.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
```

Comment:
Why not add context here, like what the request was?

Reply:
No need since it just does general error mapping to Iceberg exceptions.
Resource-specific mapping is done in method bodies if needed before this `convertExceptionIfUnsuccessful` gets called, like in getTable() for example, catching a 404 for a NoSuchTableException unlike the 404 in getDataset() which is mapped differently to a NoSuchNamespaceException.